src/include/storage/recovery/replication_log_provider.h

```diff
@@ -7,134 +7,195 @@
-  /** @return A reference to the vector of source callback IDs from the primary that should be ack'd. */
-  std::vector<uint64_t> &GetPrimaryAckables() { return primary_ackables_; }
+  /** @return True if there is an unprocessed OAT that is ready to be applied. See docs/design_replication.md. */
+  bool OATReady() const {
+    bool currently_reading_buffer = curr_buffer_ != nullptr && curr_buffer_->HasMore();
+    bool all_batches_popped = !oats_.empty() && oats_.top().batch_id_ <= last_batch_popped_;
+    return !currently_reading_buffer && all_batches_popped;
+  }
 
- private:
-  /** True if replication is currently active. */
-  bool replication_active_;
+  /**
+   * Block until a replication event is available; either replication ended, logs are received, or an OAT is available.
+   * @return The type of replication event that became available.
+   */
+  ReplicationEvent WaitUntilEvent() {
+    std::unique_lock<std::mutex> lock(replication_latch_);
+    while (true) {
+      // The order here matters. Due to the AbstractLogProvider framework that was inherited, returning LOGS without
+      // having all of the replication logs currently available may cause the log consumer to block.
+      // Moreover, check before blocking that there is really nothing to do, as cvar signals may have been missed.
+      if (!replication_active_) return ReplicationEvent::END;
+      if (OATReady()) return ReplicationEvent::OAT;
+      if (NonBlockingHasMoreRecords() || NextBatchReady()) return ReplicationEvent::LOGS;
+      replication_cv_.wait(
+          lock, [&] { return !replication_active_ || NonBlockingHasMoreRecords() || NextBatchReady() || OATReady(); });
+    }
+  }
 
-  // TODO(Gus): Put in settings manager
-  /** The number of seconds before replication times out. */
-  std::chrono::seconds replication_timeout_;
+  /**
+   * Update the latest OAT of the replication log provider.
+   *
+   * The OAT is used to signal to a replication log consumer, e.g., recovery, that all transactions up to and including
+   * OAT are now safe to be applied, as long as all replication batches up to batch ID have been processed.
+   *
+   * @param oldest_active_txn The last transaction inclusive that is safe to be applied, once batches are processed.
+   * @param batch_id The last batch ID that must have been processed before the OAT can be applied.
+   */
+  void UpdateOAT(transaction::timestamp_t oldest_active_txn, replication::record_batch_id_t batch_id) {
+    std::unique_lock lock(replication_latch_);
+    oats_.emplace(OATPair{oldest_active_txn, batch_id});
+    replication_cv_.notify_all();
+  }
 
-  // Current buffer to read logs from
-  std::unique_ptr<network::ReadBuffer> curr_buffer_ = nullptr;
+  /** @return The latest OAT to be applied, popped off the replication log. */
+  transaction::timestamp_t PopOAT() {
+    std::unique_lock lock(replication_latch_);
+    NOISEPAGE_ASSERT(!oats_.empty(), "Who's telling you to pop the OAT?");
```

Comment:
Honestly, might be better to have a more descriptive message here

Reply:
Resolved